power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -88,16 +35,216 @@ constexpr from_msgpack_t from_msgpack;
 struct from_json_t {};
 constexpr from_json_t from_json;
 
+namespace detail {
+
+// visitors for parsing
+struct DefaultNullVisitor : msgpack::null_visitor {
+    static std::string msg_for_parse_error(size_t parsed_offset, size_t error_offset, std::string_view msg) {
+        std::stringstream ss;
+        ss << msg << ", parsed_offset: " << parsed_offset << ", error_offset: " << error_offset << ".\n";
+        return ss.str();
+    }
+
+    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+    void parse_error(size_t parsed_offset, size_t error_offset) {
+        throw SerializationError{msg_for_parse_error(parsed_offset, error_offset, "Error in parsing")};
+    }
+    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+    void insufficient_bytes(size_t parsed_offset, size_t error_offset) {
+        throw SerializationError{msg_for_parse_error(parsed_offset, error_offset, "Insufficient bytes")};
+    }
+};
+
+template <class T> struct DefaultErrorVisitor : DefaultNullVisitor {
+    static constexpr std::string_view static_err_msg = "Unexpected data type!\n";
+
+    bool visit_nil() { return throw_error(); }
+    bool visit_boolean(bool /*v*/) { return throw_error(); }
+    bool visit_positive_integer(uint64_t /*v*/) { return throw_error(); }
+    bool visit_negative_integer(int64_t /*v*/) { return throw_error(); }
+    bool visit_float32(float /*v*/) { return throw_error(); }
+    bool visit_float64(double /*v*/) { return throw_error(); }
+    bool visit_str(const char* /*v*/, uint32_t /*size*/) { return throw_error(); }
+    bool visit_bin(const char* /*v*/, uint32_t /*size*/) { return throw_error(); }
+    bool visit_ext(const char* /*v*/, uint32_t /*size*/) { return throw_error(); }
+    bool start_array(uint32_t /*num_elements*/) { return throw_error(); }
```

Comment:
Use name mask is the standard way of visitors. You have some base visitor class with all the (non-virtual) visitor functions defined. If you want to have modified behavior on some of the functions, you write a derived class with definition of those functions. The defined functions in child class will mask and supersede the same function in the base class. For the functions which are not defined in child class, the base class provides the implementation.

Reply:
in that case, is it correct to derive from `msgpack::null_visitor`? if i remember correctly, the `msgpack` documentation only recommends that if the amount of overloaded functions remains small. I do think that in this implementation, we pretty much override everything, right? In that case, we can take ownership and just make things have `constexpr` implementations instead of hiding or virtual overloads

Reply:
Name masking is resolved at compile time. We have defined the `DefaultNullVisitor` which has custom parse error reporting. We inherit from that again to raise error by default for all the visitor functions. We inherit again to mask individual visitor functions for a given type.